File: src/main/java/me/arasple/mc/trmenu/hook/HookPlaceholderAPI.java
package me.arasple.mc.trmenu.hook;

import io.izzel.taboolib.internal.apache.lang3.math.NumberUtils;
import io.izzel.taboolib.module.inject.THook;
import io.izzel.taboolib.module.locale.TLocaleLoader;
import me.arasple.mc.trmenu.TrMenu;
import me.arasple.mc.trmenu.api.TrMenuAPI;
import me.arasple.mc.trmenu.utils.Querier;
import me.clip.placeholderapi.PlaceholderAPI;
import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

/**
 * @author Arasple
 * @date 2019/10/6 8:08
 * - %trmenu_tapapi_[玩家ID] | {PAPI变量}% ———— 若目标玩家在线，则返回目标玩家为请求玩家的PAPI变量
 * - %trmenu_query_[URL] | [QUERY] ---- 读取URL中JSON信息, 缓存读取中则返回 Loading... %
 * - %trmenu_query_[URL] | [QUERY] ---- 读取URL中JSON信息, 缓存读取中则返回 Loading... %
 * - %trmenu_greater_{input1}_{input2}_{containEquals}
 * - %trmenu_smaller_{input1}_{input2}_{containEquals}
 */
@THook
public class HookPlaceholderAPI extends PlaceholderExpansion {

    @Override
    public String getIdentifier() {
        return "trmenu";
    }

    @Override
    public String getAuthor() {
        return "Arasple";
    }

    @Override
    public String getVersion() {
        return TrMenu.getPlugin().getDescription().getVersion();
    }

    @Override
    public boolean persist() {
        return true;
    }


    @Override
    public String onPlaceholderRequest(Player p, String params) {
        if (p == null || !p.isOnline()) {
            return null;
        }

        String[] type = params.split("_", 2);
        String[] args = type.length >= 2 ? type[1].split("\\|") : new String[]{};
        String[] arguments = type.length >= 2 ? PlaceholderAPI.setBracketPlaceholders(p, type[1]).split("_") : new String[]{};

        if ("version".equalsIgnoreCase(type[0])) {
            return TrMenu.getPlugin().getDescription().getVersion();
        } else if ("menus".equalsIgnoreCase(type[0])) {
            return String.valueOf(TrMenuAPI.getMenus().size());
        } else if ("locale".equalsIgnoreCase(type[0])) {
            return TLocaleLoader.getLocalPriorityFirst(TrMenu.getPlugin());
        } else if (type.length == 2) {
            if (("tapapi".equalsIgnoreCase(type[0]))) {
                if (args.length == 2) {
                    Player player = Bukkit.getPlayerExact(args[0]);
                    player = player == null ? p : player;
                    return PlaceholderAPI.setBracketPlaceholders(player, args[1]);
                }
            } else if ("query".equalsIgnoreCase(type[0])) {
                if (args.length >= 1) {
                    return Querier.query(PlaceholderAPI.setBracketPlaceholders(p, args[0]), args.length >= 2 ? PlaceholderAPI.setBracketPlaceholders(p, args[1]) : null);
                }
            } else if ("greater".equalsIgnoreCase(type[0])) {
                if (arguments.length >= 2) {
                    boolean equals = arguments.length >= 3 && Boolean.parseBoolean(arguments[2]);
                    return String.valueOf(equals ? NumberUtils.toDouble(arguments[0]) >= NumberUtils.toDouble(arguments[1]) : NumberUtils.toDouble(arguments[0]) > NumberUtils.toDouble(arguments[1]));
                }
            } else if ("smaller".equalsIgnoreCase(type[0])) {
                if (arguments.length >= 2) {
                    boolean equals = arguments.length >= 3 && Boolean.parseBoolean(arguments[2]);
                    return String.valueOf(equals ? NumberUtils.toDouble(arguments[0]) <= NumberUtils.toDouble(arguments[1]) : NumberUtils.toDouble(arguments[0]) < NumberUtils.toDouble(arguments[1]));
                }
            }
        }
        return null;
    }

}

File: src/main/kotlin/me/arasple/mc/trmenu/modules/hook/HookPlaceholderAPI.kt
package me.arasple.mc.trmenu.modules.hook

import io.izzel.taboolib.internal.apache.lang3.math.NumberUtils
import io.izzel.taboolib.module.db.local.LocalPlayer
import io.izzel.taboolib.module.inject.THook
import io.izzel.taboolib.module.locale.TLocale
import io.izzel.taboolib.util.Files
import me.arasple.mc.trmenu.TrMenu
import me.arasple.mc.trmenu.data.MetaPlayer.getArguments
import me.arasple.mc.trmenu.data.MetaPlayer.getMeta
import me.arasple.mc.trmenu.data.Sessions.getMenuSession
import me.arasple.mc.trmenu.display.Menu
import me.arasple.mc.trmenu.modules.script.Scripts
import me.arasple.mc.trmenu.utils.Utils
import me.clip.placeholderapi.PlaceholderAPI
import me.clip.placeholderapi.expansion.PlaceholderExpansion
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.io.File

/**
 * @author Arasple
 * @date 2020/4/4 13:43
 */
object HookPlaceholderAPI {

    fun replace(player: OfflinePlayer, content: String): String = PlaceholderAPI.setPlaceholders(player, PlaceholderAPI.setBracketPlaceholders(player, content))

    fun replace(player: OfflinePlayer, content: List<String>): List<String> = PlaceholderAPI.setPlaceholders(player, PlaceholderAPI.setBracketPlaceholders(player, content))

    fun installDepend(): Boolean {
        val plugin = Bukkit.getPluginManager().getPlugin("PlaceholderAPI")
        val jarFile = File("plugins/PlaceholderAPI.jar")
        val url = "https://api.spiget.org/v2/resources/6245/download"
        if (plugin == null) {
            jarFile.delete()
            TLocale.sendToConsole("PLUGIN.DEPEND.DOWNLOAD", "PlaceholderAPI")
            if (Files.downloadFile(url, jarFile)) TLocale.sendToConsole("PLUGIN.DEPEND.INSTALL", "PlaceholderAPI")
            else TLocale.sendToConsole("PLUGIN.DEPEND.INSTALL-FAILED", "PlaceholderAPI")
            Bukkit.shutdown()
            return true
        }
        return false
    }

    private fun processRequest(player: OfflinePlayer, content: String): String {
        if (player !is Player) return ""
        val params = content.split("_")

        return when (params[0].toLowerCase()) {
            "menus" -> Menu.getMenus().size.toString()
            "args" -> arguments(player, params)
            "meta" -> meta(player, params)
            "data" -> if (params.size > 1) LocalPlayer.get(player).getString("TrMenu.Data.${params[1]}", "")!! else ""
            "menu" -> menu(player, params)
            "emptyslot" -> freeSlot(player, params)
            "js" -> if (params.size > 1) Scripts.script(player, params[1], true).asString() else ""
            else -> ""
        }
    }

    private fun arguments(player: Player, params: List<String>): String {
        val arguments = player.getArguments()
        if (params.size > 1) {
            return buildString {
                Utils.asIntRange(params[1]).forEach { it ->
                    arguments.getOrNull(it)?.let {
                        append(it)
                        append(" ")
                    }
                }
            }.removeSuffix(" ")
        }
        return "null"
    }

    private fun meta(player: Player, params: List<String>): String {
        return (if (params.size > 1) params[1] else null)?.let { player.getMeta(it) }?.toString() ?: "null"
    }

    private fun menu(player: Player, params: List<String>): String {
        val session = player.getMenuSession()
        return if (!session.isNull()) {
            when (params[1]) {
                "page" -> session.page.toString()
                "pages" -> session.menu?.layout?.layouts?.size.toString()
                "next" -> (session.page + 1).toString()
                "title" -> session.menu!!.settings.title.currentTitle(player)
                else -> ""
            }
        } else ""
    }

    private fun freeSlot(player: Player, params: List<String>): String {
        val session = player.getMenuSession()
        val index = NumberUtils.toInt(params.getOrElse(1) { "0" }, 0)
        val range = Utils.asIntRange(params.getOrElse(2) { "0-90" })
        return (session.menu?.getEmptySlots(player, session.page)?.filter { range.contains(it) }?.get(index) ?: -1).toString()
    }


    @THook
    class Inject : PlaceholderExpansion() {

        override fun getIdentifier() = "trmenu"

        override fun getVersion() = TrMenu.plugin.description.version

        override fun getAuthor() = "Arasple"

        override fun persist() = true

        override fun onRequest(player: OfflinePlayer?, params: String) = player?.let { processRequest(it, params) }


    }

}